### PR TITLE
Change hits file instrumentation to be memory mapped

### LINKED ISSUE #210: Spurious 'Unable to read beyond end of stream' in CI while collecting results
https://travis-ci.org/tgstation/tgstation-server/jobs/438449753#L1156

Not sure what causes this it rarely happens and seemingly at random. Latest coverlet.console

### PR DESCRIPTION
**Pull request by Cyberboss**
Fixes #210 

Added benefit of getting kernel level cleanup and atomic-ity when incrementing hit counters

### CODE REVIEW

**Comment by WiredUK** [comment]
Nice to see someone take this on, I'll try this out when I'm back at work on Monday. Thanks!

File: src/coverlet.core/Instrumentation/ModuleTrackerTemplate.cs [hunk 1]
```diff
@@ -15,130 +16,133 @@ namespace Coverlet.Core.Instrumentation
...
-                _threads.Clear();
-            }
+            this.hitsArraySize = hitsArraySize;
 
-            // The same module can be unloaded multiple times in the same process via different app domains.
-            // Use a global mutex to ensure no concurrent access.
-            using (var mutex = new Mutex(true, Path.GetFileNameWithoutExtension(HitsFilePath) + "_Mutex", out bool createdNew))
+            //now HitsFilePath and HitsArraySize should be populated
+
+            //first, try to create the hits file if it doesn't exist already
```
Comment (petli) [inline src/coverlet.core/Instrumentation/ModuleTrackerTemplate.cs]: Could you use https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile.createnew?view=netcore-2.1#System_IO_MemoryMappedFiles_MemoryMappedFile_CreateNew_System_String_System_Int64_ instead? The file doesn't actually have to contain all zeros, the OS will create zeroed memory pages as necessary.

**Comment by Cyberboss** [comment]
Think I nailed it now

**Comment by WiredUK** [comment]
> Think I nailed it now

Anecdotally at least, this works for me. Ran it a couple of times and coverage matched perfectly.

**Comment by petli** [comment]
I've gave this a spin on my code base, and it seems to count hits correctly. But it is significantly slower than version 2.3.2, where two test projects go from about 50s to respectively 90s and 135s with this change. 

I think this might be related to what @pjanotti discussed in #181 about multithreaded tests.  Even if Interlocking is less bad than using mutexes, it still adds a synchronization between the CPU cores which slows down tests. It would be interesting if @pjanotti could see what the performance is on those projects, since they seemed to be even more penalised by synchronization between parallel unit tests than the projects I test here.

This could perhaps be improved by something similar to #181 keeping one hit count memory mapped file for each CPU core in a thread-static variable and increment it without Interlocking.  But then the Coverage class must find all these files and add upp the hit counts from them.

**Comment by pjanotti** [comment]
Hi @petli, yes I tried an initial perf improvement with Interlocked and for heavy concurrency scenarios it was really bad, so had to fallback on hit counting per thread. There were some projects on the CoreFx repo that were especially bad (I can't recall exactly which ones, but the names may live in some of my previous comments). I would recommend being careful about adding interlocked in the normal hit path count if considering performance for these concurrent scenarios is relevant. Perhaps an option to use the memory mapped file?

(/cc @danmosemsft @ViktorHofer)

**Comment by petli** [comment]
I realised the performance test in coverlet can easily be adjusted to test the threaded behaviour.  This really exercises the code, since the contention gets worse if there's lots of threads all hitting the same memory pages and here's there's twenty tests all looping intensely over the same code:

    public class PerformanceTest
    {
        [Theory]
        [InlineData(20_000)]
        [InlineData(20_001)]
        [InlineData(20_002)]
        [InlineData(20_003)]
        [InlineData(20_004)]
        [InlineData(20_005)]
        [InlineData(20_006)]
        [InlineData(20_007)]
        [InlineData(20_008)]
        [InlineData(20_009)]
        [InlineData(20_010)]
        [InlineData(20_011)]
        [InlineData(20_012)]
        [InlineData(20_013)]
        [InlineData(20_014)]
        [InlineData(20_015)]
        [InlineData(20_016)]
        [InlineData(20_017)]
        [InlineData(20_018)]
        [InlineData(20_019)]
        public void TestPerformance(int iterations)

Running those 20 parallel tests on 2.3.2 took 30s on my dev machine, with the Interlocking usage in this PR it takes... \*twiddles thumbs\* ...16 minutes.  

So before this can be merged it really needs to be changed to use ThreadStatic to keep track of one memory mapped file per thread. :/

**Comment by Cyberboss** [comment]
Alright, taken care of

**Comment by petli** [comment]
I was wrong above that the test case needed additional `InlineData` to stress-test the threaded behaviour, since it already creates a lot of tasks with `Task.Run()`.  There's also a subtle bug lurking in the test which Resharper told me about and it need to change to this to give more reliable numbers:

            for (var i = 0; i < iterations; i++)
            {
                var j = i;
                tasks.Add(Task.Run(() => big.Do(j)));
            }

(I'll do a pull request on this.)

This led me to do an incorrect analysis of the test run I did before the change to thread-local files, the one I wrote about above taking 16 minutes.  It does not seem to be a problem with `Interlocking`, after all, hit counting itself is just much slower using memory mapped files rather than an `int[]`. 

Running just a single run of the performance test with 20000 iterations takes 3s with the old `int[]` counting-then-writing-to-files, 50 seconds with the pointer incrementing (either the original `Interlocking.Increment` or doing `(*hitLocationArrayOffset)++` the one-file-per-thread change). Using Read+Write instead of pointers takes 1.8 minutes.  (The test I ran with 20 `InlineData` didn't expose any additional contention, that's just 50 seconds * 20 tests = 16 minutes.)

@Cyberboss, have you seen this too or is this just something on my machine? Sorry for leading you down the wrong track here...

**Comment by Cyberboss** [comment]
I have yet to look into how to run the performance test but yeah, my own MMF experiments don't bode well with random access incrementing like this. A shared memory implementation (non-file backed) should be infinitely better, I'll see if I can get that working.

**Comment by WiredUK** [comment]
@Cyberboss Any progress on the perf testing? Keen to get this merged so we can fix our CI pipeline.

**Comment by Cyberboss** [comment]
I'll see if I can't find time to readdress this sometime this week, end of year has gotten pretty busy.

**Comment by petli** [comment]
@WiredUK @Cyberboss I've played around with this today, but just changing to a pure memory-based file didn't help much. Instead I've tried a hybrid solution where hit counting is done directly in int arrays as before, but passing the counts to the coverage analysis is done via mapped memory. That can hopefully be quicker than writing to a file and thus have time to finish running in `UnloadModule` even on loaded build servers before it the process is killed.

I have to clean up the code a bit, but will send a PR tomorrow with this.

**Comment by petli** [comment]
Finally pushed #276.

**Pull request closed without merging**